onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/base_op_builder.h

```diff
@@ -0,0 +1,90 @@
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/common/common.h"
+#include "core/providers/shared/node_unit/node_unit.h"
+#include "core/providers/nnapi/nnapi_builtin/builders/model_builder.h"
+#include "core/providers/nnapi/nnapi_builtin/builders/op_builder.h"
+#include "core/providers/nnapi/nnapi_builtin/builders/op_builder_factory.h"
+
+namespace onnxruntime {
+
+namespace common {
+class Status;
+}
+
+namespace nnapi {
+
+class ModelBuilder;
+
+template <class T>
+void CreateSharedOpBuilderImpl(const std::string& op_type,
+                               OpBuilderRegistrations& op_registrations,
+                               const std::vector<std::string>& op_types) {
+  // The shared OpSupportChecker is already in the OpSupportCheckerRegistrations
+  if (op_registrations.op_builder_map.find(op_type) != op_registrations.op_builder_map.cend())
+    return;
+
+  op_registrations.builders.push_back(std::make_unique<T>());
+  for (const auto& op : op_types) {
+    op_registrations.op_builder_map.emplace(op, op_registrations.builders.back().get());
+  }
+}
+
+class BaseOpBuilder : public IOpBuilder {
+ public:
+  virtual ~BaseOpBuilder() = default;
+
+  // Add operator related
+ public:
+  virtual void AddInitializersToSkip(ModelBuilder& /* model_builder */,
+                                     const NodeUnit& /* node_unit */) const override {}
+
+  Status AddToModelBuilder(ModelBuilder& model_builder, const NodeUnit& node_unit) const override final;
+  static bool IsOpSupported(const ModelBuilder& model_builder, const NodeUnit& node_unit);
+
+ protected:
+  virtual Status AddToModelBuilderImpl(ModelBuilder& model_builder, const NodeUnit& node_unit) const = 0;
+
+  virtual bool IsQuantizedOp(const NodeUnit& /* node_unit */) const { return false; }
+
+  // Operator support related
+ public:
+  bool IsOpSupported(const InitializedTensorSet& initializers, const NodeUnit& node_unit,
+                     const OpSupportCheckParams& params) const override;
+
+ protected:
+  virtual bool IsOpSupportedImpl(const InitializedTensorSet& /* initializers */, const NodeUnit& /* node_unit */,
+                                 const OpSupportCheckParams& /* params */) const {
+    return true;
+  }
+
+  virtual int32_t GetMinSupportedNNAPIFeatureLevel(const NodeUnit& /* node_unit */,
+                                                   const OpSupportCheckParams& /* params */) const {
+    // ANEURALNETWORKS_FEATURE_LEVEL_1 is the baseline version of NNAPI,
+    // There is no NNAPI support for Android API level 26-
+    return ANEURALNETWORKS_FEATURE_LEVEL_1;
+  }
+
+  virtual bool HasSupportedInputOutputsImpl(
+      const InitializedTensorSet& initializers, const NodeUnit& node_unit,
+      const OpSupportCheckParams& params) const;
+
+  virtual int GetMinSupportedOpSet(const NodeUnit& /* node_unit */) const { return 1; }
+  virtual int GetMaxSupportedOpSet(const NodeUnit& /* node_unit */) const { return 15; }
```

Comment:
Why is this 15? ONNX is up to opset 18 (not officially supported in this ORT release though) and I thought we supported opset 17 currently.

Reply:
sorry - accidentally updated with the old version.